ecs-agent/acs/session/task_manifest_responder.go

```diff
@@ -0,0 +1,177 @@
+// permissions and limitations under the License.
+
+package session
+
+import (
+	"sync"
+
+	"github.com/aws/amazon-ecs-agent/ecs-agent/acs/model/ecsacs"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/metrics"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/wsclient"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+type TaskManifestResponder interface {
+	New(taskComparer TaskComparer, messageIDAccessor ManifestMessageIDAccessor,
+		snAccessor SequenceNumberAccessor) wsclient.RequestResponder
+}
+
+type TaskComparer interface {
+	CompareRunningTasksOnInstanceWithManifest(*taskManifestResponder, *ecsacs.TaskManifestMessage) ([]*ecsacs.TaskIdentifier, error)
+}
+
+const (
+	TaskManifestMessageName = "TaskManifestMessage"
+)
+
+// taskManifestResponder responds to task manifest messages from ACS. It processes the
+// message and, if the message isn't stale, creates a list of stop candidates, which are
+// tasks that are running on the instance but are tasks that ACS says should not be running.
+// The taskManifestResponder will ack the task manifest message and send ACS a
+// TaskStopVerificationMessage if there are stop candidates.
+type taskManifestResponder struct {
+	taskComparer      TaskComparer
+	snAccessor        SequenceNumberAccessor
+	messageIDAccessor ManifestMessageIDAccessor
+	cluster           string
+	respond           wsclient.RespondFunc
+	logger            *logrus.Entry
+	metricsFactory    metrics.EntryFactory
+}
+
+// NewTaskManifestResponder returns an instance of the taskManifestResponder struct.
+func NewTaskManifestResponder(taskComparer TaskComparer, snAccessor SequenceNumberAccessor, messageIDAccessor ManifestMessageIDAccessor,
+	cluster string, metricsFactory metrics.EntryFactory, responseSender wsclient.RespondFunc) *taskManifestResponder {
+	r := &taskManifestResponder{
+		taskComparer:      taskComparer,
+		snAccessor:        snAccessor,
+		messageIDAccessor: messageIDAccessor,
+		cluster:           cluster,
+		logger:            logrus.WithField("ACSRequestResponder", "TaskManifestResponder"),
+		metricsFactory:    metricsFactory,
+	}
+	r.respond = ResponseToACSSender(r.Name(), responseSender)
+	return r
+}
+
+func (*taskManifestResponder) Name() string { return "task manifest responder" }
+
+func (tmr *taskManifestResponder) HandlerFunc() wsclient.RequestHandler {
+	return tmr.handleTaskManifestMessage
+}
+
+// handleTaskManifestMessage is the high level caller to handle a task manifest message from ACS.
+// It will kick off the call stack of processTaskManifestMessage calling ack and send TaskManifestMessage
+func (tmr *taskManifestResponder) handleTaskManifestMessage(message *ecsacs.TaskManifestMessage) error {
+	tmr.logger.WithField("MessageId", aws.StringValue(message.MessageId)).Info("Processing task manifest message")
+	err := tmr.processTaskManifestMessage(message)
+	tmr.metricsFactory.New(metrics.TaskManifestHandlingDuration).Done(err)
```

Comment:
Why are we using `err` to publish a duration metric?

Reply:
`err` in this context is the metric completion status. Good point about its placement though, new rev will surround `processTaskManifestMessage` with the metric creation and publishing.

Reply:
I still don't understand why the metric is named "TaskManifestResponder.Duration". What duration is referred to in the metric name?

Reply:
Discussed offline. The metric name is a legacy.